feat(reminders): handle deck deletion

ericli3690 opened the pull request:
Review reminders for a deck should be deleted when the deck they are tied to is destroyed. This PR implements this logic.

**Note that this PR is based on top of the schema migration PR, https://github.com/ankidroid/Anki-Android/pull/18856, as it heavily edits the same files. Hoping to get schema migration in quickly so this one can go in too!**

## Purpose / Description
- Added logic for checking if a deck exists before returning deck-specific reminders to ReviewRemindersDatabase. This code uses the `decks.have` method, since it seemed like the most straightforward way to accomplish deck-existence-checking.
- `have` was previously marked as "unused". I've removed this annotation.
- Added a deleteAllRemindersForDeck helper method. It's public because it will also be used in NotificationService.
- Had to mark a few database methods as suspending since they use the collection to check if a deck exists. This in turn meant wrapping some of the tests in the test file with `runTest`. Also had to explicitly create decks using `addDeck` in the test file. Moved the dummy review reminder declarations to `setUp` to accommodate this.
- Added new tests for deck deletion functionality.

## Fixes
* For GSoC 2025: Review Reminders

## Approach
- We handle deletion lazily. See the Learning section below for why. Review reminders are deleted before they must be shown or used, using the `have` method in the backend for checking if a corresponding deck exists.

## How Has This Been Tested?
- Tests pass.
- Tested on a physical Samsung S23, API 34.

## Learning (optional, can help others)
- A major problem I initially thought hard about was handling the fact that deck deletion is reversible: that is, a user can click an undo button after deleting a deck, which restores the collection to its previous form. However, this means that I cannot eagerly delete review reminders the moment the user clicks delete. At first, I came up with a complicated idea for storing a Pref of the most recently deleted decks and checking the Pref whenever the review reminders were accessed, but that idea was overly complicated.
- I then realized I could attach a callback to the deck deletion undo snackbar that would fire only if the snackbar naturally disappeared and the undo button was not pressed. However, there was still a problem with this approach: namely, that the user can cause deck deletions from other devices and cause them to appear via a sync!
- I realized I was approaching the problem completely wrong. Deck deletion should not be handled when the user clicks delete: it should be handled lazily, whenever the review reminders are actually needed. Hence I implemented this current approach of only deleting review reminders for decks if the deck is found not to exist during a review reminder access operation.

## Checklist
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

<!--- Uncomment this section ONLY if this PR introduces new resources (external libraries, icons etc)
## Licenses
_For each new external resource, add a row to the table below:_

| Library | Description | License |
| --- | --- | --- |
| Sample Icon Library | Sample Description | [The Apache Software License, Version 2.0](http://www.apache.org/licenses/LICENSE-2.0.txt) |

**Maintainers:**

* [ ] Add the https://github.com/ankidroid/Anki-Android/labels/Licenses label
* [ ] Update the [licenses](https://github.com/ankidroid/Anki-Android/wiki/Licences) wiki when merging
--->

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewRemindersDatabase.kt:
```diff
@@ -122,15 +253,42 @@ class ReviewRemindersDatabase {
+                        false // Keep this group of review reminders
+                    } else {
+                        remove(key) // Remove from SharedPreferences
+                        true // Remove from deckSpecificRemindersMap
```
No, this is an acceptable error case.

ericli3690 commented:
- Added some loggers.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewRemindersDatabase.kt:
```diff
@@ -254,15 +263,43 @@ object ReviewRemindersDatabase {
+                    if (doesDeckExist) {
+                        false // Keep this group of review reminders
+                    } else {
+                        Timber.d("Deleting review reminders for deck $did")
```
```suggestion
                        Timber.i("Deleting review reminders for deck $did")
```

ericli3690 commented:
- Changed two loggers from `.d` to `.i`.

ericli3690 commented:
Ok, I just remembered Ashish is away on vacation and probably won't be able to respond any time soon. Reading through @Scapesfear and @oyeraghib's project proposals, it seems any likely implemention will have unique collection folders for each individual profile. However, so long as each profile folder also has its own review reminders SharedPrefs file (recall that review reminders are stored in their own unique SharedPref file), this deck deletion code should still _mostly_ function as expected. In particular, if I have a review reminder for `DeckA` in `ProfileA`, and then switch to `ProfileB`, opening up ScheduleReminders should not trigger deck deletion handling code for `DeckA`'s review reminder since the code will be reading from `ProfileB`'s review reminders, which does not include the `DeckA` review reminder.

However, the code in NotificationService for firing review reminder notifications will need to be rewritten. Currently, right before firing a notification, it checks if the deck the notification is associated with still exists. If not, the notification is cancelled and the review reminder is deleted. Suppose a review reminder notification for `DeckA`'s review reminder tries to fire when `ProfileB` is active. *If* `ProfileB` does not have a deck with an identical deck ID as `DeckA`, then nothing breaks and the notification fires properly (of course, its on-click action will be broken; as previously mentioned, opening the correct collection and navigating to the correct deck will need to be implemented by the multiprofile developer). However, if an identical deck ID *does* exist in `ProfileB` (unlikely but technically possible), all review reminders for that deck will be deleted and the notification will not fire. We don't want that to happen.

That's basically all to say: the code in NotificationService will need to be rewritten to check the *corresponding* collection of a review reminder, rather than the currently active one. Then, the deck deletion handling code will need to be updated so that `withCol { decks.have(did) }` uses that corresponding collection. This means we probably want to store profile ID in ReviewReminder, which means a schema change is required.

If needed, I can implement this schema change now before review reminders are released to the public. That way, when multiprofile is released, a schema migration will not be necessary. @david-allison please let me know if I should add this placeholder field to ReviewReminder. I'm not sure about its exact data format, but @Scapesfear proposes using a UUID string and @oyeraghib proposes using the name as a string, so making a new String field should be safe. @criticalAY what's your plan for the profile ID?

Final note: currently, whenever the device boots up or the app opens, AlarmManagerService schedules all stored review reminders (with existing corresponding decks in the collection). This is because AlarmManager alarms are cleared whenever the device restarts or turns off. Once multiprofile is implemented, this logic will need to be updated so that it opens up every profile's collection one by one and schedules the corresponding review reminders - assuming, of course, that the consensus in the Discord with all review reminders from all profiles always firing is what we're implementing.

---

Otherwise, @david-allison, I don't think there's anything *directly* blocking this PR. Takeaways from both this comment and the previous one I posted above:
1. So long as unique review reminder SharedPrefs files are created for each profile, the above deck deletion code should work fine when the user is *viewing* review reminders.
2. Once multiprofile is implemented, the above deck deletion code *will* break when it comes to scheduling and sending notifications. Modifying NotificationManager and startup code to correctly check each individual profile will have to be implemented later once multiprofile is closer to being completed. Until then, all we can do is add a `profileID` field to ReviewReminder to make that code possible. Please let me know if you'd like me to implement this in a separate PR now.
3. So long as a) you're alright with switching the `AnkiDroid Directory` being a destructive operation to review reminders and b) you can't think of any other edge cases where this code breaks, I don't think there are any other changes that need to be made to this PR. If you have a different idea for how to handle `AnkiDroid Directory` switching, please let me know.

ericli3690 commented:
Update: see discussion in Discord from September 10th, 2025. The core idea is:

> Instead of deleting review reminders when their associated deck cannot be found, we can grey them out in ScheduleReminders and not fire any notifications. If the deck is found again the future (ex. due to a sync, backup restoration, AnkiDroid Directory change, etc.), the review reminder is no longer greyed out and its notifications are unmuted. When the user clicks on a greyed-out review reminder in ScheduleReminders, they can choose to reassign its deck or delete it. In other words, deleting a deck no longer immediately deletes associated review reminders: the user must manually do so from ScheduleReminders.

This will fix the problem with AnkiDroid Directory and a whole host of other issues with the current deck deletion code. It also resolves my past self's concerns here:

> Alternatively I could change the code so that if a review reminder's associated deck does not exist, we don't delete the review reminder but rather leave it hanging there and just don't display it in the UI / mute its notifications. Then, if the deck comes back into existence later, the previously-dead review reminder is reactivated. The downside here is that dead review reminders would gradually accumulate in SharedPreferences.

...since the user will be able to manually view and delete accumulating dead review reminders.

What's nice about this approach is that it also works in an acceptable way with the proposed multiprofile implementation. When a profile is switched to, so long as each profile has its own review reminders SharedPrefs file, the user will only get review reminders for that specific profile and not any others. Of course, this is not the intended way the multiprofile feature will work in its final version. According to discussion in #multiprofile, *all* profiles should have their valid review reminder notifications fired, even if the profile is not active. Hence, a multiprofile developer will need to edit NotificationService later on to make sure *all* profiles have their valid review reminder notifications fired by checking the *corresponding* profile. They will also need to edit BootService and other boot-related files to ensure *all* profiles have their review reminder notifications set after a device reboot.

As I mentioned above, I'm proposing adding a profileID field to ReviewReminder so that when a notification is about to be fired, the *corresponding* collection can be checked to see if the deck still exists. Hopefully when Ashish sees this he'll let us know what the plan is for what the profileID is going to be (uuid, string, etc.)

---

I'll work on implementing this as soon as possible. This PR is likely now dependent on https://github.com/ankidroid/Anki-Android/pull/19109. Marking this as a draft; I'll mark it as open once it's ready for review!

ericli3690 commented:
Done! Now, instead of deleting review reminders when a deck cannot be found, we simply grey the review reminder out. This means all previous concerns about actions like switching the AnkiDroid Directory or restoring from backups are moot. This does mean that dead review reminders will gradually accumulate in SharedPreferences, but since the user can view dead reminders by going to `Settings > Review Reminders`, I think this is an alright trade-off.

I've also received an answer from Ashish regarding a profileID in the ReviewReminder schema and implemented it at https://github.com/ankidroid/Anki-Android/pull/19242, so I've removed the "needs reviewer reply" tag from this PR.

Ready for review!

ericli3690 commented:
Rebased and resolved merge conflict.

david-allison reviewed (approved):
Solid solution. Thanks!

mikehardy reviewed (approved):
Does seem like a nice solution
